### PR TITLE
optimization for the sync

### PR DESCRIPTION
**Pull request by jackzhhuang**
Flexidag block synchronization improvements:

1, Optimized synchronization complexity, reduced from O(k*n) to O(n), where k is the number of block parents.
2, Reduced network requests for block synchronization.
3, Fixed reentrancy errors in reachability.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

### CODE REVIEW

File: storage/src/lib.rs [hunk 1]
```diff
@@ -168,23 +170,51 @@ static VEC_PREFIX_NAME_V4: Lazy<Vec<ColumnFamilyName>> = Lazy::new(|| {
...
     V2 = 2,
     V3 = 3,
     V4 = 4,
+    V5 = 5,
 }
 
 impl StorageVersion {
     pub fn current_version() -> StorageVersion {
-        StorageVersion::V4
+        StorageVersion::V5
```
Comment (sanlee42) [inline storage/src/lib.rs:217]: 如果是针对main network,  是则是v3->v4，是不是放v4里升了？

File: flexidag/src/reachability/inquirer.rs [hunk 2]
```diff
@@ -85,24 +85,38 @@ fn insert_to_future_covering_set(
...
+                    Ok(())
+                }
+            }
+        }
+        Err(ReachabilityError::HashesNotOrdered) => {
+            let future_covering_set = store.get_future_covering_set(merged_block)?;
+            if future_covering_set.contains(&new_block) {
+                Ok(())
+            } else {
+                Err(ReachabilityError::HashesNotOrdered)
```
Comment (sanlee42) [inline flexidag/src/reachability/inquirer.rs:116]: 这里是怎么出现的呢？如果出现了，也就是说会commit失败对吧？能否构造一个复线这个error的case

**Review by sanlee42 — CHANGES_REQUESTED** [review]

File: storage/src/lib.rs (hunk 1 above)
Comment (jackzhhuang) [reply storage/src/lib.rs:217]: 是的

File: flexidag/src/reachability/inquirer.rs (hunk 2 above)
Comment (jackzhhuang) [reply flexidag/src/reachability/inquirer.rs:116]: 多个线程很短时间内重复commit就会失败报HashesNotOrdered，这里增加一个重入的检查，产生这个失败的错误会大大降低，我连续试了10次，也不会再复现了，所以很难再构造这样的error case。

**Comment by nkysg** [comment]
前面加上dag相关描述吧

**Review by sanlee42 — APPROVED** [review]

**Pull request merged**